Dockerfile

```diff
@@ -110,21 +110,48 @@ RUN file -b   /bin/invoicer
 RUN du        /bin/invoicer
 
 
+# This stage is used to generate /etc/{group,passwd,shadow} files & avoid RUN-ing commands in the `final` layer,
+#   which would break cross-compiled images.
```

Comment:
How does `RUN` break cross-compilation?

If you know a good resource on this might be worth adding a link into this comment for others.

Reply:
To create an image for another architecture, the base image used for (at least) the very last stage has to be set to that architecture verbatim.

When `qemu` is used (ex. `lncm/bitcoind`), **all** `RUN`s in all stages are emulated.

Otherwise (ex. Go images), all but last stages can be run natively (**much** faster), and only the base of last one (`final`) has to be changed, see: https://github.com/lncm/invoicer/blob/master/scripts/ci-set-target-image-architecture.sh#L87

If I set last stage to be for an architecture that's not supported by host, ex. `FROM arm64/alpine:3.11 AS final` on an `amd64` host, and I try to `RUN` some command in it, I'll get an `exec` error, because binaries included in that base image are not able to run on the host's CPU.

Reply:
Ok that makes sense.

I'd tend to err on the side of simpler code at the cost of slower build times. I really like how you can just test locally on one arch and then wrap everything with QEMU and get portable images.

But I guess this is only a pretty minor addition so really no big deal.

Reply:
Are you recommending introducing `qemu` dependency, and setting things up to work with it, instead of copying 3 files from another simple stage?

What would the benefit be here?

I'm literally just replicating what's in `adduser` layer but in a very portable way 🤷🏻‍♂️.

Reply:
I wasn't objecting to this specific change in particular.

After you explained what's going on it makes sense but it wasn't immediately obvious to me exactly why this was needed.

I was just clarifying that as a general rule I think it's better to let QEMU handle everything. Then there's no need to work around ARCH issues like creating the user in a separate build stage or the `ci-set-target-image-architecture.sh` script overwriting the base images in the Docker file, everything is abstracted away by QEMU.

It's only one line added to the build system for QEMU and would reduce complexity elsewhere.

Reply:
> or the ci-set-target-image-architecture.sh script overwriting the base images in the Docker file

You still need to do that with `qemu`, except instead of the last one only, you need to change all `FROM`s (which can sometimes get funky if you use different base images in different stages). See: https://github.com/lncm/docker-bitcoind/blob/master/.github/workflows/on-tag.yml#L40-L56

Plus you need to take into account the fact that every command you run is being translated into a different one under-the-hood by `qemu`.  How sure are  you that `qemu` is able to do that **perfectly** for every single arbitrary assembly instruction?

Reply:
I think you could simplify that a lot with a build arg.

See:
- https://github.com/lukechilds/docker-bitcoind/blob/abd6f87ca84e24061e2ee81375db3ce6914bb11a/.github/workflows/ci.yml#L37
- https://github.com/lukechilds/docker-bitcoind/blob/abd6f87ca84e24061e2ee81375db3ce6914bb11a/Dockerfile#L4

It's one extra command in the build system to register QEMU and then a build arg to specify the architecture.

>How sure are you that qemu is able to do that perfectly for every single arbitrary assembly instruction?

It's used for emulation by the Android SDK so I'd have to imagine it's pretty accurate but I really don't know.

Reply:
That's beautiful!  Definitely going to nuke all scripts in favour of this approach 🤩